stan/math/rev/fun/dot_self.hpp

```diff
@@ -12,43 +12,42 @@
 namespace stan {
 namespace math {
 
-namespace internal {
-class dot_self_vari : public vari {
- protected:
-  vari** v_;
-  size_t size_;
-
- public:
-  dot_self_vari(vari** v, size_t size)
-      : vari(Eigen::Map<vector_vi>(v, size).val().squaredNorm()),
-        v_(v),
-        size_(size) {}
-  template <typename T, require_eigen_t<T>* = nullptr>
-  explicit dot_self_vari(const T& v)
-      : vari(v.val().squaredNorm()), size_(v.size()) {
-    v_ = reinterpret_cast<vari**>(
-        ChainableStack::instance_->memalloc_.alloc(size_ * sizeof(vari*)));
-    Eigen::Map<matrix_vi>(v_, v.rows(), v.cols()) = v.vi();
-  }
-  virtual void chain() {
-    Eigen::Map<vector_vi> v_map(v_, size_);
-    v_map.adj() += adj_ * 2.0 * v_map.val();
-  }
-};
-}  // namespace internal
-
 /**
  * Returns the dot product of a vector of var with itself.
  *
- * @tparam T type of the vector (must be derived from \c Eigen::MatrixBase and
- * have one compile time dimension equal to 1)
+ * @tparam T type of the vector (must have one compile time dimension equal to
+ * 1)
  * @param[in] v Vector.
  * @return Dot product of the vector with itself.
  */
 template <typename T, require_eigen_vector_vt<is_var, T>* = nullptr>
 inline var dot_self(const T& v) {
-  const Eigen::Ref<const plain_type_t<T>>& v_ref = v;
-  return {new internal::dot_self_vari(v_ref)};
+  const auto& v_ref = to_ref(v);
```

Comment:
Revert this (it is slower for `mat<var>`, we will put it back once we have https://github.com/stan-dev/math/pull/2144)

Reply:
I fixed this version up, just had to roll over the input once instead of twice. `custom_vari` will probs help here as well tho'

![image](https://user-images.githubusercontent.com/5857231/95647646-2282b000-0a9f-11eb-9c52-c93508872797.png)

Reply:
Oh I didn't notice this was different. This is good then.